### ISSUE TITLE
Windows 10: A dynamic link library (DLL) initialization routine failed.

### ISSUE THREAD

**Issue by McPo** [atom/node-keytar#51]
Attempting to run the application and am greeted immediately with this error.

> ELECTRON_ASAR.js:158 Uncaught Error: A dynamic link library (DLL) initialization routine failed.
> \?\C:\Users\Emmet\Sources\xxxxxxxxxxxx\src\node_modules\keytar\build\Release\keytar.node

Let me know if theres any more info needed.

**Comment by McPo** [comment]
Despite having the same Node (6.1.0) version as Electron (1.2.3). I had to install electron-rebuild and execute it on post install. I also was using electron-builder which uses two separate package.json files.

> "scripts": {
    "postinstall": "cd ../ && \"./node_modules/.bin/electron-rebuild\" -m ./src/node_modules/ -w keytar -p -f"
}

-f forces a rebuild, -p applies a relevant pre-gyp-fix and m specifies to use the inner node_modules, while obtaining the electron-prebuilt from the outer node_modules.

Still not sure why this is required though.

**Comment by gitmitch** [comment]
@McPo I'm thinking about using node-keytar for my Electron app, but this may be a blocker if I understand this issue correctly. To employ the workaround you found, it sounds like my Windows user would need Visual Studio installed because electron-rebuild would be run as the last step of the installer on their machine. Am I understanding this issue correctly?

**Comment by McPo** [comment]
I had to use electron-rebuild to get it working in Windows 10 (And I believe electron-rebuild requires Visual Studios).

However when I built it on Windows 10 it worked fine on Windows 7. So you could just rebuild the module and then store the binaries in another folder and have a post-install hook to copy your prebuilt binary over the one that its shipped with.

You might also be able to get away with just the msbuild tools, and if you need a copy of Windows to do the initial compilation. These VMs from Microsoft work well https://developer.microsoft.com/en-us/microsoft-edge/tools/vms/

Basically, it shouldn't be a blocker, just an initial inconvenience, and the tools look well across the platforms.

**Comment by endercrest** [comment]
Just wanted to confirm, the electron-rebuild solves the problem with windows 10.

**Comment by gengjiawen** [comment]
I run into this problem with Node 8.4.0, Node 7 is fine.

**Comment by manuelbieh** [comment]
I ran `electron-rebuild` 20x now with different options, I just don't get it working on Win 10 (Pro), with Fall Creators Update and Electron 1.7.8 / 1.8.1. Keep getting the `A dynamic link library (DLL) initialization routine failed` error no matter what I do ...

**Comment by acrispin** [comment]
![GitHub Logo](https://s3-us-west-2.amazonaws.com/dkads-static/other/path/my-image-test1.png)
Format: ![Alt Text](url)

**Comment by kspearrin** [comment]
@manuelbieh Did you ever get this working? I'm having same issue. `./node_modules/.bin/electron-rebuild -w keytar -p -f` doesn't seem to work.

**Comment by jarmediagmbh** [comment]
do you have electron-rebuild installed?
try to issue "`npm install electron-rebuild`" before executing "`./node_modules/.bin/electron-rebuild`"

**Comment by RyanRizzo96** [comment]
@jarmediagmbh, this solved my problem! Thanks! 💯

**Comment by TomasHubelbauer** [comment]
Can anyone please explain how does Electron Rebuild even enter the picture here? Should this module not work with bare Node, does it need Electron for something?

I am trying to use it in a VS Code extension, which is just a Node process. While VS Code itself is an Electron app, the extensions run in their own Node processes.

I am having the same issue and I have no idea how to fix this or why the fix would involve installing Electron-related stuff when I'm just trying to use this from Node.

**Comment by TomasHubelbauer** [comment]
For the record I did also struggle with statically referencing the dependency on Windows 10 and eventually just gave up to try dynamically referencing it, but also ran into issues with that (#106).

**Comment by H3npi** [comment]
When i run `electron-rebuild` the dev version works, however when i pack the application using `electron-packager` it is not working.

**Comment by buddythumbs** [comment]
I am also having this issue and am not clear from this thread how exactly to fix it, can someone post a step by step on how to fix with commands included?

Thanks

**Comment by fcapolini** [comment]
I had the same problem and published a little how-to [here](https://github.com/fcapolini/electron-keytar-windows10-howto).

Hope this helps

**Comment by fcapolini** [comment]
> I am also having this issue and am not clear from this thread how exactly to fix it, can someone post a step by step on how to fix with commands included?
> 
> Thanks

I had the same problem and published a little how-to [here](https://github.com/fcapolini/electron-keytar-windows10-howto).

Hope this helps

**Comment by shiftkey** [comment]
Closing this out because I think all that we can do currently is:

 - publish `prebuild` targets that support as much as possible (currently ever major version from Node 6 to Node 12 and Electron 1.6 to Electron 5) to save people building native modules when unnecessary
 - continue to point to the upstream [Electron docs about native Node modules](https://electronjs.org/docs/tutorial/using-native-node-modules) and [`electron-rebuild`](https://github.com/electron/electron-rebuild) docs and figure out how to make that experience better, for users who are not able to leverage `prebuild`
